File: tools/demo-smoke.mjs
// T8 demo smoke: every dist/components/*.html renders with zero real console
// errors and every IR slot name appears in the page source (live DOM or
// inside a <template>, for kernel portal content). The avatar fixture used
// to load a broken file:/// image to exercise the fallback settle; FT3
// replaced it with an inline data-URI so demos work over http — the
// fallback-on-error path stays covered by `npm run contracts -- avatar`.
// The URL filter below is legacy tolerance, kept harmless.
import { chromium } from "playwright"
import { readdirSync, readFileSync } from "node:fs"

// FT8/Step 9: dist/components now also holds per-language RTL variants
// (e.g. alert-rtl-he.html, alert-rtl-en.html) emitted by build-rtl.mjs.
// The original 49 base components are the ones that have corresponding
// IRs and slots to verify; the language variants just re-use those slots
// with translated text and don't need a separate smoke pass.
// What ships is recorded per component in src/registry/tiers.json: the three
// shipped tiers, plus `emit: true` on the four that ship despite their tier.
// demo.mjs reads the same field, so the page count here and the set it builds
// cannot disagree — this was three separate encodings of one predicate, two
// hand-written filters and a "+ 4".
const REGISTRY_TIERS = JSON.parse(readFileSync("src/registry/tiers.json", "utf8"))
const shipped = (name, tier) =>
  ["static", "kernel", "trivial-js"].includes(tier) || REGISTRY_TIERS[name]?.emit === true

const allHtml = readdirSync("dist/components").filter((f) => f.endsWith(".html")).sort()
const pages = allHtml.filter((f) => !/-rtl-(en|he|fa)\.html$/.test(f))
const rtlVariants = allHtml.length - pages.length
// base components (one per emitted IR — derived from tiers.json, not a
// hardcoded 49) + alert-demo.html (FT8/Step 1 emits this from
// examples/radix/alert-demo.tsx).
{
  const emitted = Object.entries(REGISTRY_TIERS).filter(([n, x]) => shipped(n, x.tier)).length
  if (pages.length !== emitted + 1) {
    console.error(`FAIL demo-smoke: expected ${emitted + 1} base pages (${emitted} IR + alert-demo), got ${pages.length} (${rtlVariants} RTL variants skipped)`); process.exit(1)
  }
}

// Global slot vocabulary across all emitted components — demo fixtures
// legitimately compose multiple components (e.g. label + input), so a slot
// is valid if ANY IR declares it.
const ALL_SLOTS = new Set(readdirSync("src/registry/ir")
  .map((f) => JSON.parse(readFileSync(`src/registry/ir/${f}`, "utf8")))
  .filter((ir) => shipped(ir.name, ir.tier))
  .flatMap((ir) => ir.components.flatMap((c) => c.elements.filter((e) => e.slot).map((e) => e.slot))))

const irSlots = (name) => {
  const ir = JSON.parse(readFileSync(`src/registry/ir/${name}.json`, "utf8"))
  return new Set(ir.components.flatMap((c) => c.elements.filter((e) => e.slot).map((e) => e.slot)))
}

const pageSlotNames = (html) =>
  [...new Set([...html.matchAll(/data-slot="([\w-]+)"/g)].map((m) => m[1]))]

const browser = await chromium.launch()
let fail = false
// FT8/Step 1: alert-demo.html is a composition (not a registered
// component) — it has no IR file, so skip the slot-fidelity check
// (we still render it + assert 0 console errors below).
const irExists = (name) => {
  try {
    JSON.parse(readFileSync(`src/registry/ir/${name}.json`, "utf8"))
    return true
  } catch {
    return false
  }
}
for (const f of pages) {
  const name = f.replace(".html", "")
  const html = readFileSync(`dist/components/${f}`, "utf8")
  const irSet = irExists(name) ? irSlots(name) : new Set()
  const pageSlots = pageSlotNames(html)
  // phantom check: every slot in the page must be a valid slot in some
  // emitted component (catches typos / corrupted fixtures); require ≥1 live
  // slot when this component's IR declares any.
  const phantom = pageSlots.filter((s) => !ALL_SLOTS.has(s))
  const page = await browser.newPage()
  const errors = []
  page.on("pageerror", (e) => errors.push(String(e)))
  // resource-load console messages are redundant with requestfailed (which
  // carries the URL); skip them here and let requestfailed decide.
  page.on("console", (m) => {
    if (m.type() === "error" && !/^Failed to load resource/.test(m.text()))
      errors.push(m.text())
  })
  page.on("requestfailed", (req) => {
    const url = req.url()
    if (!/nonexistent/.test(url)) errors.push(`requestfailed: ${url} ${req.failure()?.errorText || ""}`)
  })
  await page.goto(`file://${process.cwd()}/dist/components/${f}`)
  const slots = await page.evaluate(`document.querySelectorAll("[data-slot]").length`)
  await page.close()
  if (phantom.length || errors.length || (slots === 0 && irSet.size)) {
    console.error(`FAIL demo-smoke [${f}]: phantom=${phantom.join(",")} slots=${slots} irSlots=${irSet.size} errors=${JSON.stringify(errors)}`)
    fail = true
  }
}
await browser.close()
if (fail) { console.log("FAIL  demo smoke"); process.exit(1) }
console.log(`PASS  demo smoke (52 base pages + ${rtlVariants} RTL variants, IR-slot fidelity, 0 console errors)`)

File: probe/keys-go/src/demo_smoke.go
package main

// demo-smoke, ported from tools/demo-smoke.mjs — every dist/components/*.html
// renders with zero real console errors and every IR slot name appears in
// the page source. The browser half runs through the Wave-3 thin shell.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var reSlotAttr = regexp.MustCompile(`data-slot="([\w-]+)"`)
var reRtlVariantPage = regexp.MustCompile(`-rtl-(en|he|fa)\.html$`)

func runDemoSmoke() int {
	tiersB, err := os.ReadFile("src/registry/tiers.json")
	if err != nil {
		fmt.Fprintln(os.Stderr, "demo-smoke:", err)
		return 1
	}
	var regTiers map[string]struct {
		Tier string `json:"tier"`
		Emit bool   `json:"emit"`
	}
	if err := json.Unmarshal(tiersB, &regTiers); err != nil {
		fmt.Fprintln(os.Stderr, "demo-smoke: tiers:", err)
		return 1
	}
	shipped := func(name, tier string) bool {
		return tier == "static" || tier == "kernel" || tier == "trivial-js" || regTiers[name].Emit
	}

	ents, err := os.ReadDir("dist/components")
	if err != nil {
		fmt.Fprintln(os.Stderr, "demo-smoke:", err)
		return 1
	}
	var allHtml []string
	for _, e := range ents {
		if strings.HasSuffix(e.Name(), ".html") {
			allHtml = append(allHtml, e.Name())
		}
	}
	sort.Strings(allHtml)
	var pages []string
	for _, f := range allHtml {
		if !reRtlVariantPage.MatchString(f) {
			pages = append(pages, f)
		}
	}
	rtlVariants := len(allHtml) - len(pages)
	emitted := 0
	for n, t := range regTiers {
		_ = n
		if shipped(n, t.Tier) {
			emitted++
		}
	}
	if len(pages) != emitted+1 {
		fmt.Fprintf(os.Stderr, "FAIL demo-smoke: expected %d base pages (%d IR + alert-demo), got %d (%d RTL variants skipped)\n",
			emitted+1, emitted, len(pages), rtlVariants)
		return 1
	}

	// global slot vocabulary across all emitted components (fixtures
	// legitimately compose multiple components), plus each IR file's own
	// slot count — collected in the same pass so the per-page loop below
	// doesn't re-read and re-unmarshal every generated/ir/<name>.json a
	// second time.
	allSlots := map[string]bool{}
	slotCounts := map[string]int{}
	irEnts, _ := os.ReadDir("generated/ir")
	for _, e := range irEnts {
		if !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		b, _ := os.ReadFile(filepath.Join("generated/ir", e.Name()))
		var ir cssIrComponent
		if json.Unmarshal(b, &ir) != nil {
			continue
		}
		n := 0
		for _, c := range ir.Components {
			for _, el := range c.Elements {
				if el.Slot != "" {
					n++
				}
			}
		}
		slotCounts[strings.TrimSuffix(e.Name(), ".json")] = n
		if !shipped(ir.Name, ir.Tier) {
			continue
		}
		for _, c := range ir.Components {
			for _, el := range c.Elements {
				if el.Slot != "" {
					allSlots[el.Slot] = true
				}
			}
		}
	}

	shell, err := startBrowserShell()
	if err != nil {
		fmt.Fprintln(os.Stderr, "demo-smoke:", err)
		return 1
	}
	defer shell.close()
	if err := shell.launch(); err != nil {
		fmt.Fprintln(os.Stderr, "demo-smoke:", err)
		return 1
	}

	fail := false
	cwd, _ := os.Getwd()
	for _, f := range pages {
		name := strings.TrimSuffix(f, ".html")
		html, _ := os.ReadFile("dist/components/" + f)
		var phantom []string
		for _, m := range reSlotAttr.FindAllStringSubmatch(string(html), -1) {
			if !allSlots[m[1]] && !containsTok(phantom, m[1]) {
				phantom = append(phantom, m[1])
			}
		}
		nIRSlots := slotCounts[name]
		page, err := shell.newPage(true)
		if err != nil {
			fmt.Fprintln(os.Stderr, "demo-smoke:", err)
			return 1
		}
		if err := page.gotoURL("file://" + filepath.Join(cwd, "dist/components", f)); err != nil {
			fmt.Fprintln(os.Stderr, "demo-smoke:", err)
			return 1
		}
		slotsV, err := page.evaluate(`document.querySelectorAll("[data-slot]").length`)
		if err != nil {
			fmt.Fprintln(os.Stderr, "demo-smoke:", err)
			return 1
		}
		errors, _ := page.events()
		page.close()
		slots := int(slotsV.(float64))
		if len(phantom) > 0 || len(errors) > 0 || (slots == 0 && nIRSlots > 0) {
			fmt.Fprintf(os.Stderr, "FAIL demo-smoke [%s]: phantom=%s slots=%d irSlots=%d errors=%q\n",
				f, strings.Join(phantom, ","), slots, nIRSlots, errors)
			fail = true
		}
	}
	if fail {
		fmt.Println("FAIL  demo smoke")
		return 1
	}
	fmt.Printf("PASS  demo smoke (%d base pages + %d RTL variants, IR-slot fidelity, 0 console errors)\n",
		len(pages), rtlVariants)
	return 0
}
